### PR TITLE
[None][perf] Fuse MiniMax-M3 MoE routing

### PR DESCRIPTION
**Pull request by peihu-nv**
@coderabbitai summary

## Description

MiniMax-M3 MoE routing currently executes sigmoid, top-k selection, group
selection, gather, and renormalization as separate eager PyTorch operations.
TensorRT-LLM already provides the equivalent fused
`Deepseekv3RoutingImpl`/`noaux_tc_op` path.

This change reuses that fused routing implementation when the MiniMax-M3
input satisfies its CUDA, FP32, contiguous-input/bias/output contract, while
retaining the eager implementation as the fallback for unsupported inputs.
It also:

- recognizes `MiniMaxM3MoeRoutingMethod` before the parent MiniMax-M2 routing
  type in the TRTLLMGen MoE path;
- passes MiniMax-M3's routed scaling factor into the fused kernel;
- keeps generic MiniMax-M2 behavior at its existing default scale;
- preserves Step3p7's existing post-routing scaling behavior.

Unsupported inputs retain the eager path, so the functional risk is limited
to the explicitly guarded MiniMax-M3 CUDA/FP32/contiguous path.

### Performance

The original matched GB200 end-to-end A/B measured approximately 3.6% higher
serving throughput (baseline job `2422762`, candidate job `2422764`).

## Test Coverage

- Added MiniMax-M3 fused-routing coverage in
  `tests/unittest/_torch/models/test_minimax_m3.py`.
- Added MiniMax-M3 to the unified parameterized TRTLLMGen FP32-routing test in
  `tests/unittest/_torch/modules/moe/test_moe_module.py`; its reference-output
  comparison also validates the routed scaling factor.
- Current-main changed-file pre-commit checks pass.
- Exact implementation commit `f32afe87fa3156d84fa29821ca0c6937a2922006`
  compiled for SM100 on GB200 in Lyris job `2490160` (`0:0`).
- Installed-wheel GB200 validation job `2490388` passed:
  `test_minimax_m3.py -k fused_routing`: 4 passed, 27 deselected.
- Review-fix commit `62ba9ba7b600e57d3ab2b20df886a997f316cb2c`
  was validated on GB200 in job `2506508` (`0:0`):
  - annotated fused-routing tests: 4 passed, 27 deselected;
  - unified MiniMax-M3 routing matrix: NVFP4 passed; three unsupported/non-CI
    quantization combinations skipped as expected.
- Wheel SHA-256:
  `cda1ee5b9bc79030bea97457860aced13bfc59a45e684fe6e5486492c3fc380a`.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tests/unittest/_torch/modules/moe/test_moe_module.py [hunk 1]
```diff
@@ -1673,6 +1677,40 @@ def test_trtllm_gen_fp32_routing_bias(routing_method_cls, moe_model_config, quan
     )
 
 
+def test_trtllm_gen_minimax_m3_integrated_routing_scale():
```
Comment (xxi-nv) [inline tests/unittest/_torch/modules/moe/test_moe_module.py]: Do you believe this is a mandatory test?
Could this be covered by the unified test cases?

**Review by xxi-nv — APPROVED** [review]

**Review by leslie-fang25 — APPROVED** [review]

File: tests/unittest/_torch/modules/moe/test_moe_module.py (hunk 1 above)
Comment (peihu-nv) [reply tests/unittest/_torch/modules/moe/test_moe_module.py]: Good point—addressed in 62ba9ba7b6. MiniMaxM3MoeRoutingMethod is now a case in the existing parameterized test_trtllm_gen_fp32_routing_bias test, which uses the unified _test_moe_worker harness. The M3 case uses FP32 routing logits, and _create_routing_method supplies routed_scaling_factor=2.0, so the reference-output comparison preserves the integrated scale check. I removed the standalone test. GB200 job 2506508 passed the unified NVFP4 case; the other three combinations skipped under the existing CI rules as expected.

File: tests/unittest/_torch/models/test_minimax_m3.py [hunk 2]
```diff
@@ -764,6 +764,83 @@ def test_minimax_m3_routing_method_default_scale_is_identity():
     torch.testing.assert_close(same_weights, base_weights, rtol=0, atol=0)
 
 
+@pytest.mark.gpu
+@pytest.mark.skipif(not _has_cuda(), reason="fused MiniMax-M3 routing requires CUDA")
+@pytest.mark.parametrize("num_tokens", [1, 64, 8192])
+def test_minimax_m3_fused_routing_matches_reference(num_tokens, monkeypatch):
```
[the remark this answers is not in the document]
Comment (peihu-nv) [reply tests/unittest/_torch/models/test_minimax_m3.py]: Addressed in 62ba9ba7b6. Added int and pytest.MonkeyPatch parameter annotations plus -> None to test_minimax_m3_fused_routing_matches_reference, and added -> None to the CUDA Graph replay test. The third cited standalone integration test was removed after its coverage was folded into the unified parameterized MoE test. Changed-file pre-commit passes, and GB200 job 2506508 passed the updated tests.

File: tests/unittest/_torch/modules/moe/test_moe_module.py [hunk 3]
```diff
@@ -1603,6 +1607,7 @@ def _create_routing_method_with_bias(routing_method_cls, top_k, num_experts, bia
     "routing_method_cls,moe_model_config",
     [
         (MiniMaxM2MoeRoutingMethod, MoeModelConfig(256, 6, 2048, 1408)),
+        (MiniMaxM3MoeRoutingMethod, MoeModelConfig(128, 4, 512, 512)),
```
[the remark this answers is not in the document]
Comment (peihu-nv) [reply tests/unittest/_torch/modules/moe/test_moe_module.py:1610]: Thanks — I verified the current call path, and this finding does not apply. `test_trtllm_gen_fp32_routing_bias` calls `_test_moe_worker` -> `_test_moe_worker_impl` -> `_create_routing_method`, not `_create_routing_method_with_bias`. The active `_create_routing_method` factory already handles `(MiniMaxM2MoeRoutingMethod, MiniMaxM3MoeRoutingMethod)` and adds `routed_scaling_factor=2.0` for M3. `_create_routing_method_with_bias` has no callers in the test tree. The supported MiniMax-M3/NVFP4 parameterized case passed on GB200 in job `2506508`, exercising this active path. I am therefore leaving the unused helper unchanged rather than duplicating M3 logic into a dead path.

File: tensorrt_llm/_torch/modules/fused_moe/routing.py [hunk 4]
```diff
@@ -676,12 +676,33 @@ def __init__(
...
 
     def apply(
         self,
         router_logits: torch.Tensor,
         input_ids: Optional[torch.Tensor] = None,
     ) -> tuple[torch.Tensor, torch.Tensor]:
+        # M3 produces contiguous FP32 logits and bias. Keep the eager path for
+        # alternate dtype, device, shape, or layout contracts because the native
+        # op consumes raw contiguous CUDA buffers and emits FP32 weights.
+        if (router_logits.is_cuda and router_logits.dtype == torch.float32
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/modules/fused_moe/routing.py:694]: non-blocking question) Just wondering, if we are enforcing routing fusion for MinMax-M3, is there any case that falls back to original path?

**Review by yechank-nvidia — APPROVED** [review]
LGTM on modeling-side

File: tensorrt_llm/_torch/modules/fused_moe/routing.py (hunk 4 above)
Comment (peihu-nv) [reply tensorrt_llm/_torch/modules/fused_moe/routing.py:694]: Normal MiniMax-M3 serving always uses the fused path. The fallback preserves compatibility for CPU tests and unsupported tensor dtype, shape, device, or layout contracts.
